### PR TITLE
[Inductor] B2B-GEMM performance tuning with test

### PR DESCRIPTION
**Pull request by sdingcn**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130778

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by sdingcn** [comment]
This PR added
(1) performance tuning configs and
(2) dispatching criteria
for the B2B-GEMM pass. A performance test is added and passed. Here is a specific run of the performance test (the passing criterion is `average_speedup > 1`. The average is calculated using geometric mean.
```
Speedups  N = 10    N = 32    N = 60    N = 128   N = 200   
M = 256   1.866     3.646     3.587     1.019     1.034     
M = 500   1.762     2.208     1.042     1.178     0.981     
M = 1024  1.382     1.629     0.816     1.0       0.973     
M = 2000  0.967     1.041     1.003     0.997     0.987     
M = 4096  1.369     1.172     1.0       1.0       1.0       
M = 8000  2.412     1.775     1.0       0.999     1.002     
Average speedup: 1.275
```

**Comment by sdingcn** [comment]
> This PR added (1) performance tuning configs and (2) dispatching criteria for the B2B-GEMM pass. A performance test is added and passed. Here is a specific run of the performance test (the passing criterion is `average_speedup > 1`. The average is calculated using geometric mean.
> 
> ```
> Speedups  N = 10    N = 32    N = 60    N = 128   N = 200   
> M = 256   1.866     3.646     3.587     1.019     1.034     
> M = 500   1.762     2.208     1.042     1.178     0.981     
> M = 1024  1.382     1.629     0.816     1.0       0.973     
> M = 2000  0.967     1.041     1.003     0.997     0.987     
> M = 4096  1.369     1.172     1.0       1.0       1.0       
> M = 8000  2.412     1.775     1.0       0.999     1.002     
> Average speedup: 1.275
> ```

Slightly updated the dispatching criteria to use geometric mean also.
```
Speedups  N = 10    N = 32    N = 60    N = 128   N = 200   
M = 256   2.336     3.55      1.15      0.976     0.99      
M = 500   1.21      1.744     0.919     0.897     0.987     
M = 1024  1.668     1.448     1.003     1.06      0.967     
M = 2000  0.97      1.053     1.002     1.004     1.016     
M = 4096  1.356     1.157     1.0       0.999     1.0       
M = 8000  2.41      1.783     1.0       0.999     0.999     
Average speedup: 1.206
```

File: test/inductor/test_b2b_gemm.py [hunk 1]
```diff
@@ -1,40 +1,89 @@
...
         A = torch.randn((500, 100), device="cuda", dtype=torch.float16)
         B = torch.randn((100, 200), device="cuda", dtype=torch.float16)
         C = torch.randn((200, 300), device="cuda", dtype=torch.float16)
         res, (code,) = run_and_get_code(f_opt, A, B, C)
         self.assertTrue(torch.allclose(f(A, B, C), res, atol=0.2, rtol=0.01))
-        self.assertTrue("B2B_GEMM_TRITON_ENTRANCE" in code)
+        self.assertTrue("B2B_GEMM_TRITON_ENTRANCE" not in code)
+
+    @torch._dynamo.config.patch(cache_size_limit=32)
+    def test_b2b_gemm_performance(self):
```
Comment (masnesral) [inline test/inductor/test_b2b_gemm.py:43]: This looks odd to me to put a performance test in a unit test. Seems like the kind of thing that could be flakey. Did you find precedent for this elsewhere in any of our other tests?

File: test/inductor/test_b2b_gemm.py [hunk 2]
```diff
@@ -1,40 +1,89 @@
...
+        def run_with_b2b_gemm_on(m1: torch.Tensor, m2: torch.Tensor, m3: torch.Tensor) -> float:
+
+            def f(m1: torch.Tensor, m2: torch.Tensor, m3: torch.Tensor) -> torch.Tensor:
+                return torch.mm(torch.mm(m1, m2), m3)
+
+            f_opt = torch.compile(f, dynamic=False)
+            return do_bench(f_opt, (m1, m2, m3), {}, warmup = 200, rep = 2000)
+
+        speedups = []
+        print()
```
Comment (masnesral) [inline test/inductor/test_b2b_gemm.py:64]: Assuming we keep this test, I'd vote to remove `print`s. If the console output is useful, let's at least use a `log.debug`

File: torch/_inductor/fx_passes/b2b_gemm.py [hunk 3]
```diff
@@ -99,9 +105,71 @@ def b2b_gemm_grid(M, P, meta):
 )
 
 
-B2B_GEMM_PASS = PatternMatcherPass(
-    pass_name="b2b_gemm_pass",
-)
+# numbers of loads and stores in baseline and b2bgemm
```
Comment (masnesral) [inline torch/_inductor/fx_passes/b2b_gemm.py:108]: style nit: Can we use the `"""` style comments, e.g.,

```
def foo(self):
    """
    This is a function that does something.
    """
```

File: test/inductor/test_b2b_gemm.py (hunk 1 above)
Comment (sdingcn) [reply test/inductor/test_b2b_gemm.py:47]: Removed the performance assertion.

File: torch/_inductor/fx_passes/b2b_gemm.py (hunk 3 above)
Comment (sdingcn) [reply torch/_inductor/fx_passes/b2b_gemm.py]: Updated.

File: test/inductor/test_b2b_gemm.py (hunk 1 above)
Comment (eellison) [reply test/inductor/test_b2b_gemm.py:47]: Yea, DO_PERF_TEST is the precedent here.

**Review by eellison — APPROVED** [review]
nice!

**Pull request closed without merging**